pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,737 @@
+	Data   AlertData `json:"data"`
+}
+
+type AlertData struct {
+	Result []interface{} `json:"result"`
+}
+
+// TODO move to https://github.com/openshift/managed-cluster-validating-webhooks
+// validateUpgradeConfig will validate the UpgradeConfig, the desired version should be grater than or equal to the current version
+func ValidateUpgradeConfig(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (result bool, err error) {
+
+	log.Info("validating upgradeconfig")
+	clusterVersion := &configv1.ClusterVersion{}
+	err = c.Get(context.TODO(), types.NamespacedName{Name: "version"}, clusterVersion)
+	if err != nil {
+		log.Info(fmt.Sprintf("failed to get clusterversion"))
+		log.Error(err, "failed to get clusterversion")
+		return false, err
+	}
+
+	//TODO get available version from ocm api like : ocm get "https://api.openshift.com/api/clusters_mgmt/v1/versions" --parameter search="enabled='t'"
+
+	//Get current version, then compare
+	current := getCurrentlVersion(clusterVersion)
+	log.Info(fmt.Sprintf("current version is %s", current))
+	if len(current) == 0 {
+
+		return false, fmt.Errorf("failed to get current version")
+	}
+	// If the version match, it means it's already upgraded or at least control plane is upgraded.
+	if current == upgradeConfig.Spec.Desired.Version {
+		log.Info("the expected version match current version")
+		return false, fmt.Errorf("cluster is already on version %s", current)
+	}
+
+	// Compare the versions, if the current version is greater than desired, failed the validation, we don't support version rollback
+	versions := []string{current, upgradeConfig.Spec.Desired.Version}
+	log.Info("compare two versions")
+	sort.Strings(versions)
+	if versions[0] != current {
+		log.Info(fmt.Sprintf("validation failed, current version %s is greater than desired %s", current, upgradeConfig.Spec.Desired.Version))
+
+		return false, fmt.Errorf("desired version %s is greater than current version %s", upgradeConfig.Spec.Desired.Version, current)
+	}
+
+	// Find the available versions from cincinnati
+	uuid, err := uuid.Parse(string(clusterVersion.Spec.ClusterID))
+	if err != nil {
+		return false, err
+	}
+	upstreamURI, err := url.Parse(string(clusterVersion.Spec.Upstream))
+	if err != nil {
+		return false, err
+	}
+	currentVersion, err := semver.Parse(current)
+	if err != nil {
+		return false, err
+	}
+
+	updates, err := cincinnati.NewClient(uuid, nil, nil).GetUpdates(upstreamURI, runtime.GOARCH, upgradeConfig.Spec.Desired.Channel, currentVersion)
+	if err != nil {
+		return false, err
+	}
+
+	var cvoUpdates []configv1.Update
+	for _, update := range updates {
+		cvoUpdates = append(cvoUpdates, configv1.Update{
+			Version: update.Version.String(),
+			Image:   update.Image,
+		})
+	}
+	// Check whether the desired version exists in availableUpdates
+
+	found := false
+	for _, v := range cvoUpdates {
+		if v.Version == upgradeConfig.Spec.Desired.Version && v.Force == false {
```

Comment:
I'm curious why the "vForce == false" here? Not objecting to it, just want to understand why.

Reply:
> 	// force allows an administrator to update to an image that has failed
> 	// verification, does not appear in the availableUpdates list, or otherwise
> 	// would be blocked by normal protections on update. This option should only
> 	// be used when the authenticity of the provided image has been verified out
> 	// of band because the provided image will run with full administrative access
> 	// to the cluster. Do not use this flag with images that comes from unknown
> 	// or potentially malicious sources.
> 	//
> 	// This flag does not override other forms of consistency checking that are
> 	// required before a new update is deployed.

This is the description for the `False` congifuration.

Reply:
Thanks, that now makes perfect sense. Cheers!